Implement `HttpMatcher::and` and `HttpMatcher::or`

dmackdev opened the pull request:
Addresses part of  https://github.com/plabayo/rama/issues/92 and https://github.com/plabayo/rama/issues/93 for `http` layer.

- Modifies `HttpMatcherKind::All` and `HttpMatcherKind::Any` variants to contain `Vec<HttpMatcher>` instead of `Vec<HttpKind>`. This way the negation of each matcher is preserved, and this fixes the bug I mentioned in the above issue.
- Implements `HttpMatcher::and` and `HttpMatcher::or` methods.
- Updates existing `HttpMatcher::and_*` and `HttpMatcher::or_*` methods to use new methods.
- Adds `HttpKind::Custom(_)` variant for custom matcher implementations.
- Adds tests.

GlenDC commented on src/http/matcher/mod.rs:
```diff
@@ -883,3 +605,131 @@ impl<State, Body> crate::service::Matcher<State, Request<Body>> for HttpMatcherK
+            .uri("www.example.com")
+            .body(())
+            .unwrap();
+        assert!(!matcher.matches(None, &Context::default(), &req));
```
for these combo tests might be sane to verify that the logic is indeed correct (e.g. in this expect we expect to behave it like `!(A&B&C)`. Currently however this kind of test will pass even when it is `A&B&!C`.

This can be achieved by adding more request checks for the same combo matcher.

GlenDC reviewed (commented):
Looks already very good so far @dmackdev ! Well done.
I agree with your comment in https://github.com/plabayo/rama/issues/93#issuecomment-2027700683, if you feel motivated to pick that one up as well, feel free to take it as yours, and you can handle it as part of the same PR if you want. Because indeed, using request-agnostic booleans it will be indeed be a lot easier to test an entire truth table :)

All yours if you want. Do let it be known in that issue in the case.

Currently your code looks excellent for me and in fact you already solved the ticket for what is me concerned. Only remark is that currently you are just testing a single case of an entire truth table. I would prefer it that for each of these tests we did cover the entire truth table as expected. Especially for the complex queries.

GlenDC commented on src/http/matcher/mod.rs:
```diff
@@ -880,6 +696,309 @@ impl<State, Body> crate::service::Matcher<State, Request<Body>> for HttpMatcherK
+    #[test]
+    fn test_matcher_negation_and_header_contains() {
+        let matcher = HttpMatcher::method_get().negate().and_header_contains(
+            "content-type".parse().unwrap(),
```
You can use `HeaderName/HeaderValue`::from_static for these cases, just make sure to keep the names lower case as otherwise you get a panic, but that's okay for your cases.

GlenDC commented:
Your PR looks good to me @dmackdev , thank you for all your hard work.

Do you still plan to add some abstract truth table tests with hardcoded bools. Given your current test coverage I am honestly fine without, but might be the cherry on the cake if you do still go for it. Either way fine for me.

dmackdev commented:
Thank you @GlenDC , and thank you for the assistance along the way.

I was planning on amending the tests I added to use the hardcoded bool custom matchers - are you suggesting to add more tests instead of amending the ones I added?

I can make the changes this evening.

GlenDC commented on src/http/matcher/mod.rs:
```diff
@@ -880,6 +696,504 @@ impl<State, Body> crate::service::Matcher<State, Request<Body>> for HttpMatcherK
+
+    #[test]
+    fn test_matcher_ands_combination() {
+        let cases = [
```
You can add `itertools` as a dev dependency.
This way you can just do:

```rust
use itertools::Itertools;

#[test]
fn test_matcher_ands_combination() {
    for (a, b, c) in [true, false].permutations(3) {
        let expected = a && b && c;
        let a = BooleanMatcher(a);
        let b = BooleanMatcher(b);
        let c = BooleanMatcher(c);

        let matcher = a.and(b).and(c);
        let req = Request::builder().body(()).unwrap();
        assert_eq!(
            matcher.matches(None, &Context::default(), &req),
            expected,
            "({:#?}).matches({:#?})",
            matcher,
            req
        );
    }
}
```

Seems a lot less trouble, and easier to read/maintain?
If you agree you can apply this for all these tests.

GlenDC commented on src/http/matcher/mod.rs:
```diff
@@ -880,6 +696,517 @@ impl<State, Body> crate::service::Matcher<State, Request<Body>> for HttpMatcherK
+    }
+
+    #[test]
+    fn test_matcher_or_and_or_and_negation() {
```
I would just apply the same technique I propose in my other comment. This way you can easily test it without having to write the cases yourself :) And easier to reason about then these manual http cases. So I would suggest to replace them, but only if you apply a technique such as the ones approached, because manual test cases won't make it that much easier.

GlenDC reviewed (changes requested):
Once the last comments are applied, let me know if it is finished from your POV. Because then I'll just merge it in also next time I review, in case I approve at that point.

dmackdev commented:
@GlenDC all done from my side.

GlenDC reviewed (approved):
Great job. Thank you for this. Merging it in.
Take care and feel free to take on another open issue if you can and wish.
Either way thanks for your devotion and work.
